addons/mail/static/src/model/model_manager.js

```diff
@@ -1143,13 +1380,29 @@ export class ModelManager {
      * @returns {boolean} whether any value changed for the current record
      */
     _update(record, data, options = {}) {
+        this._ensureNoLockingListener();
+        const disabler = ['_update', record, data, options];
+        this._listenerDisablers.add(disabler);
         if (!record.exists()) {
             throw Error(`Cannot update already deleted record ${record.localId}.`);
         }
         const { allowWriteReadonly = false } = options;
         const Model = record.constructor;
         let hasChanged = false;
-        for (const fieldName of Object.keys(data)) {
+        const sortedFieldNames = Object.keys(data);
+        sortedFieldNames.sort((a, b) => {
+            // Always update identifying fields first because updating other relational field will
+            // trigger update of inverse field, which will require this identifying fields to be set
+            // beforehand to properly detect whether this is already linked or not on the inverse.
+            if (Model.__identifyingFieldsFlattened.has(a) && !Model.__identifyingFieldsFlattened.has(b)) {
+                return -1;
+            }
+            if (!Model.__identifyingFieldsFlattened.has(a) && Model.__identifyingFieldsFlattened.has(b)) {
+                return 1;
+            }
+            return 0;
+        });
```

Comment:
What part of it do you think could be slow?

`sortedFieldNames` will usually contain a very low amount of elements (at most the number of fields on a record, but typically much less, between 1 and 5).

Even if that wasn't the case, as long as the compare function is fast, native `sort` should be fast even on large data set.
It only uses O(1) operations inside the sort method.

It's a good concern to have, but in this case I think it would be premature optimization to change anything, as the code is not obviously wrong (?), and no issue is noticeable in practice.

Reply:
I was just concerned by using `.sort()` in every update cycle if it's used a lot with many items. Keep in mind `.sort()` uses bubble sort, which is easy-to-use and highly customisable, but it doesn't scale good.